src/openrct2/core/JobPool.hpp

```diff
@@ -0,0 +1,176 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * A full copy of the GNU General Public License can be found in licence.txt
+ *****************************************************************************/
+#pragma endregion
+
+#pragma once
+
+#include <atomic>
+#include <cassert>
+#include <condition_variable>
+#include <deque>
+#include <functional>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+class JobPool
+{
+private:
+    struct TaskData
+    {
+        const std::function<void()> WorkFn;
+        const std::function<void()> CompletionFn;
+
+        TaskData(std::function<void()> workFn, std::function<void()> completionFn)
+            : WorkFn(workFn),
+              CompletionFn(completionFn)
+        {
+        }
+    };
+
+    std::atomic_bool _shouldStop = { false };
+    std::atomic<size_t> _processing = { 0 };
+    std::vector<std::thread> _threads;
+    std::deque<TaskData> _pending;
+    std::deque<TaskData> _completed;
+    std::condition_variable _condPending;
+    std::condition_variable _condComplete;
+    std::mutex _mutex;
+
+    typedef std::unique_lock<std::mutex> unique_lock;
+
+public:
+    JobPool()
+    {
+        for (size_t n = 0; n < std::thread::hardware_concurrency(); n++)
+        {
+            _threads.emplace_back(&JobPool::ProcessQueue, this);
+        }
+    }
+
+    ~JobPool()
+    {
+        {
+            unique_lock lock(_mutex);
+            _shouldStop = true;
+            _condPending.notify_all();
+        }
+
+        for (auto&& th : _threads)
+        {
+            assert(th.joinable() != false);
+            th.join();
+        }
+    }
+
+    void AddTask(std::function<void()> workFn, std::function<void()> completionFn)
+    {
+        unique_lock lock(_mutex);
+        _pending.emplace_back(workFn, completionFn);
+        _condPending.notify_one();
+    }
+
+    void AddTask(std::function<void()> workFn)
+    {
+        return AddTask(workFn, nullptr);
+    }
+
+    void Join(std::function<void()> reportFn = nullptr)
+    {
+        unique_lock lock(_mutex);
+        while (true)
+        {
+            // Wait for the queue to become empty or having completed tasks.
+            _condComplete.wait(lock, [this]()
+            {
+                return (_pending.empty() && _processing == 0) ||
+                        !_completed.empty();
+            });
+
+            // Dispatch all completion callbacks if there are any.
+            while (!_completed.empty())
+            {
+                auto taskData = _completed.front();
+                _completed.pop_front();
+
+                if (taskData.CompletionFn)
+                {
+                    lock.unlock();
+
+                    taskData.CompletionFn();
+
+                    lock.lock();
+                }
+            }
+
+            if (reportFn)
+            {
+                lock.unlock();
+
+                reportFn();
```

Comment:
In case of this PR, we pass `reportProgress` here which _simply_ prints out to `stdout`. I'm a bit unhappy this is handled without any locks **and** there is no comment stating whose responsibility it is to handle locking.